File: Programming and Projection Methods/Lab11/Laborator11Hibernate/CompanyClient/src/main/java/company/client/Controller.java
package company.client;

import company.model.*;
import company.services.AppException;
import company.services.IObserver;
import company.services.IServer;

import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;


import java.io.Serializable;
import java.rmi.RemoteException;
import java.rmi.server.UnicastRemoteObject;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class Controller extends UnicastRemoteObject implements IObserver, Serializable {
    private IServer server;
    private User user;
    private ObservableList<Ride> rModel;
    private List<Pair<ObservableList<RBooking>, List<String>>> rbModelList;

    public Controller(IServer serv) throws RemoteException {
        this.server = serv;
        rModel = FXCollections.observableArrayList();
        rbModelList = new ArrayList<>();
    }

    private void populateRideList() {
        rModel.clear();
        Ride[] rides = server.findAllRides();
        for (Ride r : rides) {
            rModel.add(r);
        }
    }

    public ObservableList<Ride> getRModel() {
        populateRideList();
        return rModel;
    }

    public ObservableList<RBooking> getRBModel(String destination, String date, String hour) {
        ObservableList<RBooking> rbModel = FXCollections.observableArrayList();
        RBooking[] rbs = server.findAllRBookings(destination, date, hour);
        for (RBooking rb : rbs) {
            rbModel.add(rb);
        }
        rbModelList.add(new Pair<>(rbModel, Arrays.asList(destination, date, hour)));
        return rbModel;
    }

    public void findUser(String username, String password) {
        User user = new User(username, password);
        server.login(user, this);
        this.user = user;
    }

    public void logout(User u) {
        server.logout(u, this);
    }

    public String bookPlaces(Ride r, String cname, int nrplaces) {
        return server.bookPlaces(r, new Client(0, cname), nrplaces);
    }

    private void upd(Pair<ObservableList<RBooking>, List<String>> pair) {
        pair.getKey().clear();
        RBooking[] rbs = server.findAllRBookings(pair.getValue().get(0), pair.getValue().get(1), pair.getValue().get(2));
        for (RBooking rb : rbs) {
            pair.getKey().add(rb);
        }
    }

    @Override
    public void update() throws AppException {
        Platform.runLater(new Runnable() {
            @Override
            public void run() {
                populateRideList();
                rbModelList.forEach(x -> upd(x));
            }
        });
    }
}

File: Programming and Projection Methods/Lab10/Laborator10CSharp/CompanyClient/Controller.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading;
using CompanyMode;
using CompanyModel;
using Thrift.Server;
using Thrift.Transport;

namespace CompanyClient
{
    public class Controller : IObserver.Iface
    {
        public event EventHandler<UserEventArgs> UpdateEvent;
        private IServer.Iface server;
        private User user;
        private int port;
        private IList<Object> rModel;
        private IList<KeyValuePair<IList<RBooking>, IList<String>>> rbModelList;

        public Controller(IServer.Iface server)
        {
            this.server = server;
            rModel = new List<Object>();
            rbModelList = new List<KeyValuePair<IList<RBooking>, IList<String>>>();
        }

        public User FindUser(String username, String password)
        {
            User user = new User(username, password);
            int p = server.login(user);
            this.user = user;
            this.port = p;
            new Thread(ClientServ).Start();
            return user;
        }

        public void ClientServ()
        {
            IObserver.Processor processor = new IObserver.Processor(this);
            TServerTransport serverTransport = new TServerSocket(port);
            TServer serverc = new TThreadPoolServer(processor, serverTransport);
            serverc.Serve();
        }


        private void PopulateRideList()
        {
            rModel.Clear();
            IList<Ride> rides = server.findAllRides();
            foreach (var r in rides)
                rModel.Add(new
                {
                    Destinatia = r.Destination,
                    Data = r.Date,
                    Ora = r.Hour,
                    NrLocuriDisponibile = r.NrPlacesAvailable
                });
        }

        public IList<Object> GetRModel()
        {
            PopulateRideList();
            return rModel;
        }

        public IList<RBooking> GetRBModel(string destination, string date, string hour)
        {
            IList<RBooking> rbModel = new ObservableCollection<RBooking>();
            server.findAllRBookings(destination, date, hour).ToList().ForEach(x => rbModel.Add(x));
            rbModelList.Add(new KeyValuePair<IList<RBooking>, IList<string>>(rbModel, new List<string> { destination, date, hour }));
            return rbModel;
        }


        public void Logout(User user)
        {
            server.logout(user);
        }

        public String BookPlaces(Ride r, String cname, int nrplaces)
        {
            String places = server.bookPlaces(r, new Clientj(0, cname), nrplaces);
            return places;
        }

        private void Upd(KeyValuePair<IList<RBooking>, IList<string>> pair)
        {
            pair.Key.Clear();
            server.findAllRBookings(pair.Value.ElementAt(0), pair.Value.ElementAt(1), pair.Value.ElementAt(2))
                .ToList().ForEach(x => pair.Key.Add(x));
        }

        protected virtual void OnUserEvent(UserEventArgs e)
        {
            if (UpdateEvent == null)
                return;
            UpdateEvent(this, e);
            Console.WriteLine("Update event called");
        }

        public void update()
        {
            Console.WriteLine("Controller update");
            UserEventArgs userArgs = new UserEventArgs();
            OnUserEvent(userArgs);
        }
    }
}